app/services/proxy_deployment_service.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+class ProxyDeploymentService
+  delegate :apicast_configuration_driven,
+           :deployable?,
+           :service_mesh_integration?,
+           :provider,
+           :proxy_configs, to: :@proxy
+
+  class UnknownStageError < ArgumentError; end
+
+  def self.call(*args)
+    new(*args).call
+  end
+
+  def initialize(proxy, environment: :staging, v1_compatible: false)
+    @proxy = proxy
+    @environment = environment
+    @v1_compatible = v1_compatible
+  end
+
+  def call
+    case @environment
+    when :staging, :sandbox
+      deploy
+    when :production
+      deploy_production
+    else
+      raise UnknownStageError, "Unknown environment: #{@stage}"
+    end
+  end
+
+  private
+
+  def deploy
+    return true unless deployable?
+
+    if service_mesh_integration?
+      deploy_v2 && deploy_production_v2
+    elsif apicast_configuration_driven
+      deploy_v2
+    elsif @v1_compatible
+      deploy_v1
+    end
+  end
+
+  # TODO: in order to include this method into 'call', do I need another service?
+  def deploy_production
+    if apicast_configuration_driven
+      deploy_production_v2
+    elsif @proxy.ready_to_deploy?
+      provider.deploy_production_apicast
+    end
+  end
+
+  # Deprecated
+  def deploy_v1
+    ProxyDeploymentV1Service.new(provider).deploy(@proxy)
```

Comment:
It's now clear IMO that the name of this service should have been `ApicastV1DeploymentService`, instead of `ProxyDeploymentV1Service `.